core/consensus/src/engine.rs

```diff
@@ -199,6 +201,8 @@ impl<Adapter: ConsensusAdapter + 'static> Engine<FixedPill, FixedSignedTxs>
             )
             .await?;
 
+        let mut set = self.exemption_hash.write();
```

Comment:
There are potential deadlocks and you need to release the locks as soon as possible.

A good practice is to encapsulate clear as a function.